### PR TITLE
[Serve] Add method info to HTTP metrics

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

This change updates some metrics that Serve uses to gauge HTTP requests successes and failures:

* `serve_num_http_requests`
    * Adds the `method` field
* `serve_num_http_error_requests`
    * Adds the `method` field
* `serve_num_deployment_http_error_requests`
    * Adds the `error_code` field
    * Adds the `method` field

This helps distinguish between different methods used when pinging the same endpoint.

Link to documentation changes: https://ray--29918.org.readthedocs.build/en/29918/serve/production-guide/monitoring.html#built-in-ray-serve-metrics

## Related issue number

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - This change adds a unit test to `test_metrics.py`.
   - Manually verified that these metrics show up in the output using this script:

```python
# File name: repro.py
import ray
from ray import serve

import logging
import requests
from fastapi import FastAPI

ray.init(address="auto")

app = FastAPI()
logger = logging.getLogger("ray.serve")

@serve.deployment(route_prefix="/hello")
@serve.ingress(app)
class MyFastAPIDeployment:

    def __init__(self):
        logger.setLevel(logging.DEBUG)

    @app.get("/")
    def root(self):
        logger.info("What a wonderful kind of day!")
        return "Hello, world!"


serve.run(MyFastAPIDeployment.bind())
resp = requests.get("http://localhost:8000/hello")
assert resp.json() == "Hello, world!"
```

**Repro steps:**
1. `ray start --head --metrics-export-port=8080`
2. `python repro.py`
3. Go to `localhost:8080`. After a short period of time, the updated metrics should appear with the `GET` method and error code `307`.

### CODE REVIEW

**Comment by simon-mo** [comment]
add test to https://github.com/ray-project/ray/blob/5f84b7d07bca487c82fd4cf048071176728c7e50/python/ray/serve/tests/test_metrics.py#L70 ?

**Review by simon-mo — APPROVED** [review]
LGTM pending tests

**Comment by shrekris-anyscale** [comment]
@simon-mo I added a unit test to check for the new fields.

**Review by simon-mo — APPROVED** [review]

**Review by edoakes — APPROVED** [review]

**Review by sihanwang41 — APPROVED** [review]

File: python/ray/serve/_private/http_proxy.py [hunk 1]
```diff
@@ -322,10 +339,18 @@ async def __call__(self, scope, receive, send):
         )
         if status_code != "200":
             self.request_error_counter.inc(
-                tags={"route": route_path, "error_code": status_code}
+                tags={
+                    "route": route_path,
+                    "error_code": status_code,
+                    "method": scope["method"].upper(),
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/http_proxy.py:345]: I added `upper()` there to match the behavior of [`access_log_message`](https://github.com/ray-project/ray/blob/e5a8b1dd55817305ac85237cd99d8b6ed23df294/python/ray/serve/_private/logging_utils.py#L12-L14), which [we currently use when logging the request](https://github.com/ray-project/ray/blob/efee158cecc49fbec5527e75e17faadba4fac48d/python/ray/serve/_private/http_proxy.py#L315-L322). I couldn't find any documentation from `uvicorn` guaranteeing that the method would always be uppercase, so I believe we need it for correctness.

**Pull request merged**